### PR TITLE
refactor: remove branch node pool

### PR DESCRIPTION
**Pull request by rphmeier**
This now uses the `PagePool` to store branch pages. Outdated branches will be naturally dropped at the end of `finish_sync`, when the index is replaced with the new version.

I changed the index to store `Arc<BranchNode>` where `BranchNode` is a thin wrapper around a `FatPage`. The Arcs will carry some overhead. This overhead currently doesn't show up in profiling. 

Of course, there is also a viable path to simply storing `Page`s, unsafely dereferencing them, and manually dropping them at the end of `finish_sync`.

### CODE REVIEW

**Comment by rphmeier** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/398?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#398** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/398?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#392** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/392?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#393](https://github.com/thrumdev/nomt/pull/393) <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/393?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#391** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/391?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#382** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/382?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#381** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/381?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/beatree/ops/reconstruction.rs [hunk 1]
```diff
@@ -9,19 +9,20 @@
...
     index::Index,
 };
+use crate::io::PagePool;
 
 /// Reconstruct the upper branch nodes of the btree from the bottom branch nodes and the leaf nodes.
 /// This places all branches into the BNP and returns an index into all BBNs.
 pub fn reconstruct(
     bn_fd: File,
-    bnp: &mut branch::BranchNodePool,
+    page_pool: &PagePool,
```
Comment (pepyakin) [inline nomt/src/beatree/ops/reconstruction.rs:25]: Should update the references to BNP I suppose?

File: nomt/src/beatree/bbn.rs [hunk 2]
```diff
@@ -36,13 +36,18 @@ pub fn open(
 }
 
 impl BbnStoreWriter {
-    pub fn allocate(&mut self, mut branch_node: BranchNode) -> PageNumber {
+    /// Allocate a PN for the given branch.
```
Comment (pepyakin) [inline nomt/src/beatree/bbn.rs:39]: Should mention that the branch node is assigned to that PN!

**Review by pepyakin — APPROVED** [review]

**Comment by rphmeier** [comment]
### Merge activity

* **Sep 6, 7:06 PM EDT**: @rphmeier started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thrumdev/nomt/398).

**Pull request merged**